File: 811 - Subdomain Visit Count.java
import java.util.*;


class Solution {
    public List<String> subdomainVisits(String[] cpdomains) {
        Map<String, Integer> m = new HashMap<>();
        for (String domains : cpdomains) {
            String[] countdomain = domains.split("\\s+");
            int count = Integer.valueOf(countdomain[0]);
            m.put(countdomain[1], m.getOrDefault(countdomain[1], 0) + count);
            String[] domain = countdomain[1].split("\\.");
            String temp = "";
            for (int i = domain.length - 1; i > 0; i--) {
                temp = i == domain.length - 1 ? domain[i] : domain[i] + "." + temp;
                String joinedDomain = temp;
                m.put(joinedDomain, m.getOrDefault(joinedDomain, 0) + count);
            }
        }
        
        List<String> res = new ArrayList<>();
        int j = 0;
        for (Map.Entry<String, Integer> i : m.entrySet()) {
            res.add(String.valueOf(i.getValue()) + " " + i.getKey());
        }
        return res;
    }
}

File: 811 - Subdomain Visit Count.py
class Solution:
    def subdomainVisits(self, cpdomains: list[str]) -> list[str]:
        counts = dict()
        for c in cpdomains:
            c = c.split()
            allc = int(c[0])
            counts[c[1]] = counts.setdefault(c[1], 0) + allc
            c[1] = c[1].split('.')
            comb = [c[1][i:] for i in range(-(len(c[1]) - 1), 0)]
            for item in comb:
                item = '.'.join(item)
                counts[item] = counts.setdefault(item, 0) + allc
        res = []
        for i in counts.items():
            temp = ''
            temp += str(i[1])
            temp += ' '
            temp += i[0]
            res.append(temp)
        return res


s = Solution()
a = ["900 google.mail.com", "50 yahoo.com", "1 intel.mail.com", "5 wiki.org"]
print(s.subdomainVisits(a))
